datanommer.models/datanommer/models/__init__.py

```diff
@@ -128,7 +130,16 @@ def add(envelope):
     obj.msg = message['msg']
     obj.headers = headers
 
-    session.add(obj)
+    # Construct an INSERT ... SELECT ... which will do nothing if an
+    # entry with a matching msg_id already exists. This can happen in the
+    # event of message redelivery.
+    cols = [c for c in Message.__table__.c if c.name != 'id']
+    vals = [literal(getattr(obj, c.name)) for c in cols]
+    not_exists = ~exists(select([Message.__table__.c.i]).\
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```

Comment:
> If it's possible to avoid triggering an error condition, it seems like it's sensible to do so.

I don't agree with this approach as I feel it breaks with a very common Python pattern. Error conditions exist to be triggered. It's far better to handle them explicitly than to introduce unnecessary complexity to dodge them.

> The other case is two workers processing two different messages, but which both relate to the same user or package. The get_or_create() method had a race condition between checking for the existence of that user/package and inserting it. When there was a collision, the second worker would hit a unique constraint violation, and the message would get lost.

To me, this is a signal that the transaction doesn't make sense. Nested transactions or a series of transactions seems appropriate.